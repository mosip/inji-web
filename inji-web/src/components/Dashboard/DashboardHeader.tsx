import React, {useState, useEffect, useRef} from "react";
import {useTranslation} from "react-i18next";
import {useNavigate} from "react-router-dom";
import {LanguageSelector} from "../Common/LanguageSelector";
import {useSelector} from "react-redux";
import {RootState} from "../../types/redux";
import {api} from "../../utils/api";
import {useCookies} from "react-cookie";
import {toast} from "react-toastify";
import {RiArrowDownSFill, RiArrowUpSFill} from "react-icons/ri";
import {GradientWrapper} from "../Common/GradientWrapper";
import { convertStringIntoPascalCase } from "../../pages/Dashboard/utils";

export const DashboardHeader: React.FC = () => {
    const language = useSelector((state: RootState) => state.common.language);
    const {t} = useTranslation("Dashboard");
    const navigate = useNavigate();
    const [displayName, setDisplayName] = useState<string | null>(null);
    const [isProfileDropdownOpen, setIsProfileDropdownOpen] = useState(false);
    const [cookies] = useCookies(["XSRF-TOKEN"]);
    const dropdownRef = useRef<HTMLDivElement>(null);

    useEffect(() => {
        setDisplayName(localStorage.getItem("displayName"));
    }, [localStorage.getItem("displayName")]);

    useEffect(() => {
        const handleClickOutside = (event: MouseEvent) => {
            if (
                dropdownRef.current &&
                !dropdownRef.current.contains(event.target as Node)
            ) {
                setIsProfileDropdownOpen(false);
            }
        };

        document.addEventListener("mousedown", handleClickOutside);
        return () => {
            document.removeEventListener("mousedown", handleClickOutside);
        };
    }, []);

    const handleLogout = async () => {
        try {
            const apiRequest = api.userLogout;
            const response = await fetch(apiRequest.url(), {
                method: "POST",
                credentials: "include",
                headers: {
                    "X-XSRF-TOKEN": cookies["XSRF-TOKEN"]
                }
            });

            if (response.ok) {
                localStorage.removeItem("displayName");
                localStorage.removeItem("walletId");
                window.location.replace("/");
            } else {
                const parsedResponse = await response.json();
                const errorCode = parsedResponse?.errors[0].errorCode;
                if (errorCode === "user_logout_error") {
                    localStorage.removeItem("displayName");
                    window.location.replace("/login");
                }
                throw new Error(parsedResponse?.errors[0]?.errorMessage);
            }
        } catch (error) {
            console.error("Logout failed with error:", error);
            toast.error("Unable to logout. Please try again.");
        }
    };

    const dropdownItems: DropdownItem[] = [
        {
            label: "Profile",
            onClick: () => navigate("profile"),
            textColor: "text-gray-700"
        },
        {
            label: "FAQ",
            onClick: () =>
                navigate("faq", {state: {from: window.location.pathname}}),
            textColor: "text-gray-700"
        },
        {
            label: "Logout",
            onClick: handleLogout,
            textColor: "text-red-700"
        }
    ];

    const toggleProfileDropdown = () => {
        setIsProfileDropdownOpen(!isProfileDropdownOpen);
    };

    return (
        <header className="bg-transparent">
            <div className="container mx-auto px-4 py-4 flex justify-between items-center">
                <div className="flex items-center">
                    <img
                        src={require("../../assets/InjiWebLogo.png")}
                        className="h-13 w-28 scale-150 cursor-pointer"
                        alt="Inji Web Logo"
                        onClick={() => navigate("/")}
                    />
                </div>

                <div className="flex items-center space-x-6">
                    <LanguageSelector />

                    <div className="relative" ref={dropdownRef}>
                        <div
                            className="flex items-center space-x-2 cursor-pointer"
                            onClick={toggleProfileDropdown}
                        >
                            <div className="w-10 h-10 rounded-full bg-gradient-to-r from-iw-primary to-iw-secondary flex items-center justify-center text-white font-medium text-sm">
                                {displayName
                                    ? displayName
                                          .split(" ")
                                          .map((name) =>
                                              name.charAt(0).toUpperCase()
                                          )
                                          .join("")
                                    : "U"}
                            </div>
                            <span className="font-medium text-gray-800">
                                {convertStringIntoPascalCase(displayName)}
                            </span>

                            {/* Dropdown arrow */}
                            <div className="relative inline-block">
                                {isProfileDropdownOpen ? (
                                    <GradientWrapper>
                                        <RiArrowUpSFill
                                            size={20}
                                            color={
                                                "var(--iw-color-languageArrowIcon)"
                                            }
                                        />
                                    </GradientWrapper>
                                ) : (
                                    <GradientWrapper>
                                        <RiArrowDownSFill
                                            size={20}
                                            color={
                                                "var(--iw-color-languageArrowIcon)"
                                            }
                                        />
                                    </GradientWrapper>
                                )}
                            </div>
                        </div>

                        {/* Profile dropdown */}
                        {isProfileDropdownOpen && (
                            <div className="absolute -right-7 top-100 mt-8 w-56 bg-white rounded-lg shadow-lg z-50">
                                <div className="absolute top-[-0.3rem] right-8 w-3 h-3 bg-white transform rotate-45" />
                                <div className="py-2">
                                    {dropdownItems.map((item, index) => (
                                        <React.Fragment key={index}>
                                            <div
                                                className={`px-4 py-2 text-sm ${item.textColor} hover:bg-gray-100 cursor-pointer`}
                                                onClick={item.onClick}
                                            >
                                                {item.label}
                                            </div>
                                            {index !==
                                                dropdownItems.length - 1 && (
                                                <hr className="border-gray-200 my-1 mx-2" />
                                            )}
                                        </React.Fragment>
                                    ))}
                                </div>
                            </div>
                        )}
                    </div>
                </div>
            </div>
        </header>
    );
};

type DropdownItem = {
    label: string;
    onClick: () => void;
    textColor: string;
};
