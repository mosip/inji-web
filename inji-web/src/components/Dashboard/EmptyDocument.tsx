import React from "react";
import {useTranslation} from "react-i18next";
import {useNavigate} from "react-router-dom";

export const EmptyDocument: React.FC = () => {
    const {t} = useTranslation("Dashboard");
    const navigate = useNavigate();

    return (
        <div className="container mx-auto sm:px-2 md:px-4 lg:px-6 py-6 relative ml-3 sm:ml-0">
            <div className="flex flex-col sm:flex-row justify-between items-start mb-4 sm:mb-6 gap-4 sm:gap-0">
                <div className="flex items-start">
                    <div className="flex items-start">
                        <svg
                            data-testid={'Back-Arrow-Icon'}
                            width="29"
                            height="29"
                            viewBox="0 0 24 18"
                            fill="none"
                            xmlns="http://www.w3.org/2000/svg"
                            className="mr-2 cursor-pointer"
                            onClick={() => navigate('/dashboard/home')}
                        >
                            <path
                                d="M20 11H7.83l5.59-5.59L12 4l-8 8 8 8 1.41-1.41L7.83 13H20v-2z"
                                fill="#000000"
                            />
                        </svg>
                    </div>
                    <div className="flex flex-col items-start">
                        <span
                            data-testid={'Stored-Credentials'}
                            className="text-2xl font-medium"
                        >
                            {t('StoredCredentials.title')}
                        </span>
                        <span
                            data-testid={'Home'}
                            className="text-xs sm:text-sm text-[#5B03AD] cursor-pointer hover:underline"
                            onClick={() => navigate('/dashboard/home')}
                        >
                            {t('Home.title')}
                        </span>
                    </div>
                </div>
                <div className="hidden sm:block">
                    <button
                        data-testid="Add-Credential"
                        onClick={() => navigate('/dashboard/home')}
                        className="bg-gradient-to-r from-[#FF5300] via-[#C5363C] to-[#5B03AD] text-white px-6 py-2 rounded-lg text-sm font-semibold shadow-sm transition-shadow"
                    >
                        {t('StoredCredentials.header.addCredential')}
                    </button>
                </div>
            </div>

            <div
                data-testid={'Blank-Document'}
                className="bg-white rounded-lg shadow-iw-emptyDocuments p-6 sm:p-8 md:p-10 flex flex-col items-center justify-center min-h-[500px] my-4 sm:my-8 md:my-16"
            >
                <svg
                    data-testid={'Document-Icon'}
                    width="88"
                    height="88"
                    viewBox="0 0 68 88"
                    fill="none"
                    xmlns="http://www.w3.org/2000/svg"
                    className="mb-8 sm:mb-12 md:mb-20"
                >
                    <path
                        d="M17.847 88C15.611 88 13.744 87.246 12.246 85.739C10.749 84.231 10 82.353 10 80.102V7.898C10 5.647 10.749 3.769 12.246 2.261C13.744 0.754 15.611 0 17.847 0H56.143L78 22V80.102C78 82.353 77.251 84.231 75.754 85.739C74.256 87.246 72.389 88 70.153 88H17.847ZM53.714 24.444H73.143L53.714 4.889V24.444ZM27 70.889H61V66H27V70.889ZM27 51.333H61V46.444H27V51.333Z"
                        fill="#DEDEDE"
                    />
                </svg>
                <h2
                    data-testid={'No-Credentials-Title'}
                    className="text-xl text-center sm:text-2xl font-medium text-gray-800 mb-2"
                >
                    {t('StoredCredentials.emptyScreen.title')}
                </h2>
                <p
                    data-testid={'No-Credentials-Subtitle'}
                    className="text-gray-500 text-center max-w-md font-small"
                >
                    Lorem ipsum dolor sit amet consectetur. Ut nis facilisi
                    condimentum odio nunc sem nisi.
                </p>
                <div className="block sm:hidden mt-6">
                    <button
                        data-testid={'Blank-Document-Add-Credential'}
                        onClick={() => navigate('/dashboard/home')}
                        className="bg-gradient-to-r from-[#FF5300] via-[#C5363C] to-[#5B03AD] text-white px-6 py-2 rounded-lg text-sm font-semibold shadow-sm transition-shadow"
                    >
                        {t('StoredCredentials.header.addCredential')}
                    </button>
                </div>
            </div>
        </div>
    );
};
