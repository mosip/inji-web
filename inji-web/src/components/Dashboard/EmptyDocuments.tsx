import React from "react";
import { useTranslation } from "react-i18next";
import { useNavigate } from "react-router-dom";
import backgroundImage from "../../assets/Background.svg";

export const EmptyDocuments: React.FC = () => {
  const { t } = useTranslation("Dashboard");
  const navigate = useNavigate();

  return (
    <div className="min-h-screen bg-[#F9FAFB] relative">
      <div className="container mx-auto px-4 py-6 relative z-10">
        <div className="flex flex-col sm:flex-row justify-between items-start sm:items-center mb-4 sm:mb-6 gap-4 sm:gap-0">
          <div className="flex items-center">
            <div className="flex items-center">
              <svg width="24" height="24" viewBox="0 0 24 24" fill="none" xmlns="http://www.w3.org/2000/svg" className="mr-2 cursor-pointer" onClick={() => navigate("/")}>
                <path d="M20 11H7.83l5.59-5.59L12 4l-8 8 8 8 1.41-1.41L7.83 13H20v-2z" fill="#000000"/>
              </svg>
              <h1 className="text-xl sm:text-2xl font-semibold">Stored Credentials</h1>
            </div>
            <span className="text-xs sm:text-sm text-[#5B03AD] ml-2 cursor-pointer hover:underline" onClick={() => navigate("/")}>Home</span>
          </div>
          <button
            onClick={() => navigate("/issuers")}
            className="bg-gradient-to-r from-[#FF5300] via-[#C5363C] to-[#5B03AD] text-white px-4 py-2 rounded-lg text-sm font-semibold shadow-sm hover:shadow-md transition-shadow"
          >
            Add Credential
          </button>
        </div>

        <div className="bg-white rounded-lg shadow-md p-6 sm:p-8 md:p-10 flex flex-col items-center justify-center min-h-[300px]">
          <svg width="68" height="88" viewBox="0 0 68 88" fill="none" xmlns="http://www.w3.org/2000/svg" className="mb-6">
            <path d="M17.847 88C15.611 88 13.744 87.246 12.246 85.739C10.749 84.231 10 82.353 10 80.102V7.898C10 5.647 10.749 3.769 12.246 2.261C13.744 0.754 15.611 0 17.847 0H56.143L78 22V80.102C78 82.353 77.251 84.231 75.754 85.739C74.256 87.246 72.389 88 70.153 88H17.847ZM53.714 24.444H73.143L53.714 4.889V24.444ZM27 70.889H61V66H27V70.889ZM27 51.333H61V46.444H27V51.333Z" fill="#DEDEDE"/>
          </svg>
          <h2 className="text-xl font-medium text-gray-800 mb-2">
            No documents stored yet
          </h2>
          <p className="text-gray-500 text-center max-w-md">
            Lorem ipsum dolor sit amet consectetur. Ut nis facilisi condimentum odio nunc sem nisi.
          </p>
        </div>
      </div>

      <div className="absolute inset-0 z-0 overflow-hidden">
        <img src={backgroundImage} alt="Background" className="absolute top-0 left-0 w-full h-full object-cover" />
      </div>
    </div>
  );
};
