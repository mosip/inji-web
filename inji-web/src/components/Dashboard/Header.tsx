import React, {useState, useEffect, useRef, forwardRef} from 'react';
import {useNavigate} from 'react-router-dom';
import {LanguageSelector} from '../Common/LanguageSelector';
import {api} from '../../utils/api';
import {useCookies} from 'react-cookie';
import {toast} from 'react-toastify';
import {RiArrowDownSFill, RiArrowUpSFill} from 'react-icons/ri';
import {GradientWrapper} from '../Common/GradientWrapper';
import {
    convertStringIntoPascalCase,
    getProfileInitials
} from '../../pages/Dashboard/utils';
import {useUser} from '../../hooks/useUser';
import {DropdownItem} from './types';
import HamburgerMenu from '../../assets/HamburgerMenu.svg';

export const Header = forwardRef<HTMLDivElement, any>((props, ref) => {
    const navigate = useNavigate();
    const [displayName, setDisplayName] = useState<string | undefined>(
        undefined
    );
    const [isProfileDropdownOpen, setIsProfileDropdownOpen] = useState(false);
    const [isHamburgerMenuOpen, setIsHamburgerMenuOpen] = useState(false);
    const [cookies] = useCookies(['XSRF-TOKEN']);
    const dropdownRef = useRef<HTMLDivElement>(null);
    const hamburgerMenuRef = useRef<HTMLImageElement>(null);
    const {user, removeUser} = useUser();
    const displayNameFromLocalStorage = user?.displayName;
    const hasProfilePictureUrl = user?.profilePictureUrl;

    useEffect(() => {
        setDisplayName(displayNameFromLocalStorage);
    }, [displayNameFromLocalStorage]);

    useEffect(() => {
        const handleClickOutside = (event: MouseEvent) => {
            const target = event.target as Node;

            if (dropdownRef.current && !dropdownRef.current.contains(target)) {
                setIsProfileDropdownOpen(false);
            }

            if (
                hamburgerMenuRef.current &&
                !hamburgerMenuRef.current.contains(target)
            ) {
                setIsHamburgerMenuOpen(false);
            }
        };

        document.addEventListener('mousedown', handleClickOutside);
        return () =>
            document.removeEventListener('mousedown', handleClickOutside);
    }, []);

    const handleLogout = async () => {
        try {
            const apiRequest = api.userLogout;
            const response = await fetch(apiRequest.url(), {
                method: 'POST',
                credentials: 'include',
                headers: {
                    'X-XSRF-TOKEN': cookies['XSRF-TOKEN']
                }
            });

            if (response.ok) {
                removeUser();
                localStorage.removeItem('walletId');
                window.location.replace('/');
            } else {
                const parsedResponse = await response.json();
                const errorCode = parsedResponse?.errors[0].errorCode;
                if (errorCode === 'user_logout_error') {
                    removeUser();
                    window.location.replace('/login');
                }
                throw new Error(parsedResponse?.errors[0]?.errorMessage);
            }
        } catch (error) {
            console.error('Logout failed with error:', error);
            toast.error('Unable to logout. Please try again.');
        }
    };

    const dropdownItems: DropdownItem[] = [
        {
            label: 'Profile',
            onClick: () => {
                setIsProfileDropdownOpen(false);
                navigate('profile');
            },
            textColor: 'text-gray-700'
        },
        {
            label: 'FAQ',
            onClick: () => {
                setIsProfileDropdownOpen(false);
                navigate('faq', {state: {from: window.location.pathname}});
            },
            textColor: 'text-gray-700'
        },
        {
            label: 'Logout',
            onClick: handleLogout,
            textColor: 'text-red-700'
        }
    ];

    const toggleProfileDropdown = () => {
        setIsProfileDropdownOpen(!isProfileDropdownOpen);
    };

    const toggleHamburgerMenu = () => {
        setIsHamburgerMenuOpen(!isHamburgerMenuOpen);
    };

    const getUserProfileIconWithName = () => (
        <div className="flex gap-2 items-center">
            <div
                className={`aspect-square w-12 sm:w-14 rounded-full bg-[#DDDDDD] overflow-hidden flex items-center justify-center text-[#1A001D] font-medium text-lg ${
                    !hasProfilePictureUrl && 'p-2 sm:p-3 md:p-4'
                }`}
            >
                {hasProfilePictureUrl ? (
                    <img
                        src={user.profilePictureUrl}
                        alt="Profile Pic"
                        className="w-full h-full object-cover rounded-full"
                        referrerPolicy="no-referrer"
                    />
                ) : (
                    getProfileInitials(displayName)
                )}
            </div>
            <span className="font-semibold text-gray-800">
                {convertStringIntoPascalCase(displayName)}
            </span>
        </div>
    );

    return (
        <header
            ref={ref}
            data-testid="Dashboard-Header-Container"
            className="fixed top-0 left-0 right-0 z-20 bg-iw-background bg-transparent shadow-[0_4px_5px_0_rgba(0,0,0,0.051)]"
        >
            <div className="container mx-auto px-4 py-4 flex justify-between items-center">
                <div className="flex items-center gap-7">
                    <div
                        data-testid="Hamburger-Menu"
                        className="block sm:hidden w-full"
                    >
                        <img
                            data-testid="Hamburger-Menu-icon"
                            ref={hamburgerMenuRef}
                            src={HamburgerMenu}
                            alt="Hamburger Menu"
                            onClick={toggleHamburgerMenu}
                            className="cursor-pointer"
                        />
                    </div>
                    <img
                        data-testid="Inji-Wallet-Logo"
                        src={require('../../assets/InjiWebLogo.png')}
                        className="h-13 w-28 scale-150 cursor-pointer"
                        alt="Inji Web Logo"
                        onClick={() => navigate('/')}
                    />
                </div>

                <div className="flex items-center space-x-6">
                    <LanguageSelector />

                    <div
                        data-testid="Profile-Details"
                        className="hidden sm:block relative"
                        ref={dropdownRef}
                    >
                        <div className="flex items-center space-x-2">
                            {getUserProfileIconWithName()}
                            <div
                                className="relative inline-block cursor-pointer"
                                onClick={toggleProfileDropdown}
                            >
                                {isProfileDropdownOpen ? (
                                    <GradientWrapper>
                                        <RiArrowUpSFill
                                            size={20}
                                            color={
                                                'var(--iw-color-languageArrowIcon)'
                                            }
                                        />
                                    </GradientWrapper>
                                ) : (
                                    <GradientWrapper>
                                        <RiArrowDownSFill
                                            size={20}
                                            color={
                                                'var(--iw-color-languageArrowIcon)'
                                            }
                                        />
                                    </GradientWrapper>
                                )}
                            </div>
                        </div>

                        {isProfileDropdownOpen && (
                            <div
                                data-testid="Profile-Dropdown"
                                className="absolute -right-7 top-100 mt-8 w-56 bg-white rounded-lg shadow-lg z-50 font-medium"
                            >
                                <div className="absolute top-[-0.3rem] right-8 w-3 h-3 bg-white transform rotate-45" />
                                <div className="py-2">
                                    {dropdownItems.map((item, index) => (
                                        <React.Fragment key={index}>
                                            <div
                                                className={`px-4 py-2 text-sm ${item.textColor} hover:bg-gray-100 cursor-pointer`}
                                                onClick={item.onClick}
                                            >
                                                {item.label}
                                            </div>
                                            {index !==
                                                dropdownItems.length - 1 && (
                                                <hr className="border-gray-200 my-1 mx-2" />
                                            )}
                                        </React.Fragment>
                                    ))}
                                </div>
                            </div>
                        )}
                    </div>
                </div>
            </div>
            <div
                data-testid="Hamburger-Menu-dropdown"
                className="block sm:hidden w-full"
            >
                {isHamburgerMenuOpen && (
                    <div
                        style={{marginTop: props.headerHeight}}
                        className="absolute top-1 bg-white shadow-iw-hamburger-dropdown p-2 w-full"
                    >
                        <div>
                            <div className="flex items-center px-4 py-2">
                                {getUserProfileIconWithName()}
                            </div>
                            {dropdownItems.map((item, index) => (
                                <React.Fragment key={index}>
                                    <div
                                        className={`px-4 py-2 text-sm ${item.textColor} hover:bg-gray-100 cursor-pointer font-medium`}
                                        onClick={item.onClick}
                                    >
                                        {item.label}
                                    </div>
                                    {index !== dropdownItems.length - 1 && (
                                        <hr className="border-gray-200 my-1 mx-2" />
                                    )}
                                </React.Fragment>
                            ))}
                        </div>
                    </div>
                )}
            </div>
        </header>
    );
});

Header.displayName = 'DashboardHeader';
