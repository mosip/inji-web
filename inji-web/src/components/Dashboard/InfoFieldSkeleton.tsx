import { InfoFieldSkeletonProps } from "./types";
export const InfoFieldSkeleton = ({
    width = 'w-full',
    height = 'h-4',
    className = '',
  }: InfoFieldSkeletonProps) => {
    return (
      <div
        className={`bg-gray-300 rounded animate-pulse ${width} ${height} ${className}`}
      />
    );
  };
  