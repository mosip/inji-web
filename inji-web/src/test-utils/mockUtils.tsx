import React, { ReactElement } from 'react';
import { render, RenderOptions } from '@testing-library/react';
import { Provider } from 'react-redux';
import { BrowserRouter, Route, BrowserRouter as Router, Routes } from 'react-router-dom';
import { reduxStore } from '../redux/reduxStore';

// Mock for storage module
export const mockStorageModule = () => {
    jest.mock('../utils/storage.ts', () => ({
        storage: {
            getItem: jest.fn(),
            setItem: jest.fn(),
            SESSION_INFO: 'SESSION_INFO',
            SELECTED_LANGUAGE: 'selectedLanguage'
        }
    }));
};

// Mock for localStorage
export const mockLocalStorage = () => {
    let store: { [key: string]: string } = {};
  
    const localStorageMock = {
      getItem: jest.fn((key: string) => store[key] || null),
      setItem: jest.fn((key: string, value: string) => {
        store[key] = value;
      })
    };
  
    Object.defineProperty(window, 'localStorage', { 
      value: localStorageMock,
      writable: true
    });
  
    return localStorageMock;
  };

// API Mock
export const mockApi = {
    authorizationRedirectionUrl: 'http://mockurl.com'
};

// Crypto Mock
export const mockCrypto = {
    getRandomValues: (array: Uint32Array) => {
        for (let i = 0; i < array.length; i++) {
            array[i] = 4294967296;
        }
        return array;
    },
    subtle: {} as SubtleCrypto,
    randomUUID: () => '123e4567-e89b-12d3-a456-426614174000'
} as Crypto;

export const mockUseTranslation = () => {
    jest.mock('react-i18next', () => ({
        useTranslation: () => ({
            t: (key: string) => key,
        }),
    }));
};

export const mockUseNavigate = () => {
    const mockNavigate = jest.fn();
    jest.mock('react-router-dom', () => ({
        ...jest.requireActual('react-router-dom'),
        useNavigate: mockNavigate,
    }));
};

export const mockUseSelector = () => {
    jest.mock('react-redux', () => ({
        ...jest.requireActual('react-redux'),
        useSelector: jest.fn(),
    }));
};

export const setMockUseSelectorState = (state: any) => {
    const useSelectorMock = require('react-redux').useSelector;
    useSelectorMock.mockImplementation((selector: any) => selector(state));
};

export const mockUseGetObjectForCurrentLanguage = () => {
    jest.mock('../utils/i18n', () => ({
        getObjectForCurrentLanguage: jest.fn(),
    }));
};

export const wrapUnderRouter = (children: React.ReactNode) => {
    return <Router>{children}</Router>;
};

export const mockUseSearchCredentials = () => {
    jest.mock('../components/Credentials/SearchCredential', () => ({
        SearchCredential: () => <></>,
    }));
 };

export const mockUseDispatch = () => {
    jest.mock('react-redux', () => ({
        ...jest.requireActual('react-redux'),
        useDispatch: jest.fn(),
    }));
};
              
export const mockUseSpinningLoader = () => {
    jest.mock('../components/Common/SpinningLoader', () => ({
        SpinningLoader: () => <></>,
    }));
};

export const mockUseLanguageSelector = () => {
    jest.mock('../components/Common/LanguageSelector', () => ({
        LanguageSelector: () => <></>,
    }));
};

interface RenderWithProviderOptions extends Omit<RenderOptions, 'queries'> {}

export const renderWithProvider = (element: ReactElement, options?: RenderWithProviderOptions) => {
    return render(
        <Provider store={reduxStore}>
            <Router>
                {element}
            </Router>
        </Provider>,
        options
    );
};
    
export const mockUseParams = ()=>{
    jest.mock('react-router-dom', () => ({
        ...jest.requireActual('react-router-dom'),
        useParams: () => ({ issuerId: 'test-issuer-id' }),
      }));   
};

export const mockUseapiObject = () =>{
    jest.mock('../utils/api', () => ({
        api: {
          mimotoHost: 'http://mocked-api-host',
        },
      }));
};

export const mockUseFetch = () =>{
    jest.mock('../hooks/useFetch', () => {
        const RequestStatus = {
          LOADING: 'LOADING',
          DONE: 'DONE',
          ERROR: 'ERROR',
        };
        return {
          useFetch: () => jest.fn(),
          RequestStatus,
        };
      });
};

export const mockUseToast = () =>{
    jest.mock('react-toastify', () => ({
      toast: {
        error: jest.fn(),
      },
    }));
};

export const mockusemisc = ()=>{
    jest.mock('../utils/misc', () => ({
        downloadCredentialPDF: jest.fn(),
        getErrorObject: jest.fn(),
        getTokenRequestBody: jest.fn(),
      }));
};

export const mockusei18n = ()=>{
    jest.mock('react-i18next', () => ({
        useTranslation: () => ({
          t: (key: string) => key,
        }),
        initReactI18next: {
          type: '3rdParty',
          init: jest.fn(),
        },
      }));

};

export const mockWindowLocation = (url: string) => {
    const location = new URL(url) as unknown as Location;
    Object.defineProperty(window, 'location', {
      value: location,
      writable: true,
    });
  };

export const renderWithRouter = (Element: React.ReactElement, { route = '/' } = {}) => {
    window.history.pushState({}, 'Test page', route);
    return render(
      <BrowserRouter>
        <Provider store={reduxStore}>
          <Routes>
            <Route path="*" element={Element} />
          </Routes>
        </Provider>
      </BrowserRouter>
    );
  };

