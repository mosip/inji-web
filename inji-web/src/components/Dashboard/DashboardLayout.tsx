import React from "react";
import {Sidebar} from "./Sidebar";
import {DashboardHeader} from "./DashboardHeader";
import {Footer} from "../PageTemplate/Footer";
import {useSelector} from "react-redux";
import {RootState} from "../../types/redux";
import {getDirCurrentLanguage} from "../../utils/i18n";
import {Outlet} from "react-router-dom";

export const DashboardLayout: React.FC = () => {
    const language = useSelector((state: RootState) => state.common.language);

    return (
        <div
            className="h-screen flex flex-col bg-iw-background font-base overflow-hidden"
            dir={getDirCurrentLanguage(language)}
        >
            <DashboardHeader />

            <div className="flex flex-1 overflow-hidden">
                <Sidebar />
                <main className="flex-1 bg overflow-y-auto -mt-1">
                    <div className="container mx-auto px-4 sm:px-6 lg:px-8 py-6">
                        <Outlet />
                    </div>
                </main>
            </div>
            <Footer />
        </div>
    );
};
