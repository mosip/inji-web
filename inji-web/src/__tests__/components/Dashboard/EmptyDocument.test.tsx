import {render, screen, fireEvent} from '@testing-library/react';
import {EmptyDocument} from '../../../components/Dashboard/EmptyDocument';
import {mockusei18n, renderWithRouter} from '../../../test-utils/mockUtils';

jest.mock('react-i18next', () => ({
    useTranslation: () => ({
        t: (key: string) => key
    }),
    initReactI18next: {
        type: '3rdParty',
        init: jest.fn()
    }
}));

describe('Testing the Layout and Functionalities of EmptyDocumentPage ->', () => {
    let renderedComponent: ReturnType<typeof render>;

    beforeEach(() => {
        jest.clearAllMocks();
        mockusei18n();
        renderedComponent = renderWithRouter(<EmptyDocument />);
    });

    it('check if the layout is matching with snapshot', () => {
        const {asFragment} = renderedComponent;

        expect(asFragment()).toMatchSnapshot();
    });

    it('check if it renders text elements correctly using translation keys', () => {
        expect(screen.getByTestId('Stored-Credentials')).toBeInTheDocument();
        expect(screen.getByTestId('Home')).toBeInTheDocument();
        expect(screen.getByTestId('Back-Arrow-Icon')).toBeInTheDocument();
        expect(screen.getByTestId('Add-Credential')).toBeInTheDocument();
        expect(screen.getByTestId('Blank-Document')).toBeInTheDocument();
        expect(screen.getByTestId('Document-Icon')).toBeInTheDocument();
        expect(screen.getByTestId('No-Credentials-Title')).toBeInTheDocument();
    });

    it('should navigate to dashboard home on nav back button click', () => {
        fireEvent.click(screen.getByTestId('Back-Arrow-Icon'));

        expect(window.location.pathname).toBe('/dashboard/home');
    });

    it('should navigate to dashboard home on Home text click', () => {
        fireEvent.click(screen.getByTestId('Home'));

        expect(window.location.pathname).toBe('/dashboard/home');
    });

    it('should navigate to dashboard home when Add credential button is clicked in larger screens', () => {
        setScreenWidth(764);

        fireEvent.click(screen.getByTestId('Add-Credential'));

        expect(window.location.pathname).toBe('/dashboard/home');
    });

    it('should navigate to dashboard home when Add credential button in blank document is clicked in smaller screens', () => {
        setScreenWidth(400);

        fireEvent.click(screen.getByTestId('Add-Credential'));

        expect(window.location.pathname).toBe('/dashboard/home');
    });
});

const setScreenWidth = (width: number) => {
    window.innerWidth = width;
    window.dispatchEvent(new Event('resize'));
};
