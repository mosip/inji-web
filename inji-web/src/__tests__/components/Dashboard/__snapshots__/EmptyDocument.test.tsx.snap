// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing the Layout and Functionalities of EmptyDocument component check if the layout is matching with snapshot 1`] = `
<DocumentFragment>
  <div
    class="container mx-auto sm:px-2 md:px-4 lg:px-6 py-6 relative ml-3 sm:ml-0"
  >
    <div
      class="flex flex-col sm:flex-row justify-between items-start mb-4 sm:mb-6 gap-4 sm:gap-0"
    >
      <div
        class="flex items-start"
      >
        <div
          class="flex items-start"
        >
          <svg
            class="mr-2 cursor-pointer"
            data-testid="Back-Arrow-Icon"
            fill="none"
            height="29"
            viewBox="0 0 24 18"
            width="29"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M20 11H7.83l5.59-5.59L12 4l-8 8 8 8 1.41-1.41L7.83 13H20v-2z"
              fill="#000000"
            />
          </svg>
        </div>
        <div
          class="flex flex-col items-start"
        >
          <span
            class="text-2xl font-medium"
            data-testid="Stored-Credentials"
          >
            StoredCredentials.title
          </span>
          <span
            class="text-xs sm:text-sm text-[#5B03AD] cursor-pointer hover:underline"
            data-testid="Home"
          >
            Home.title
          </span>
        </div>
      </div>
      <div
        class="hidden sm:block"
      >
        <button
          class="bg-gradient-to-r from-[#FF5300] via-[#C5363C] to-[#5B03AD] text-white px-6 py-2 rounded-lg text-sm font-semibold shadow-sm transition-shadow"
          data-testid="Add-Credential"
        >
          StoredCredentials.header.addCredential
        </button>
      </div>
    </div>
    <div
      class="bg-white rounded-lg shadow-iw-emptyDocuments p-6 sm:p-8 md:p-10 flex flex-col items-center justify-center min-h-[500px] my-4 sm:my-8 md:my-16"
      data-testid="Blank-Document"
    >
      <svg
        class="mb-8 sm:mb-12 md:mb-20"
        data-testid="Document-Icon"
        fill="none"
        height="88"
        viewBox="0 0 68 88"
        width="88"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M17.847 88C15.611 88 13.744 87.246 12.246 85.739C10.749 84.231 10 82.353 10 80.102V7.898C10 5.647 10.749 3.769 12.246 2.261C13.744 0.754 15.611 0 17.847 0H56.143L78 22V80.102C78 82.353 77.251 84.231 75.754 85.739C74.256 87.246 72.389 88 70.153 88H17.847ZM53.714 24.444H73.143L53.714 4.889V24.444ZM27 70.889H61V66H27V70.889ZM27 51.333H61V46.444H27V51.333Z"
          fill="#DEDEDE"
        />
      </svg>
      <h2
        class="text-xl text-center sm:text-2xl font-medium text-gray-800 mb-2"
        data-testid="No-Credentials-Title"
      >
        StoredCredentials.emptyScreen.title
      </h2>
      <p
        class="text-gray-500 text-center max-w-md font-small"
        data-testid="No-Credentials-Subtitle"
      >
        Lorem ipsum dolor sit amet consectetur. Ut nis facilisi condimentum odio nunc sem nisi.
      </p>
      <div
        class="block sm:hidden mt-6"
      >
        <button
          class="bg-gradient-to-r from-[#FF5300] via-[#C5363C] to-[#5B03AD] text-white px-6 py-2 rounded-lg text-sm font-semibold shadow-sm transition-shadow"
          data-testid="Blank-Document-Add-Credential"
        >
          StoredCredentials.header.addCredential
        </button>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
